<a href="https://colab.research.google.com/github/iambrookedrake/DS-Unit-1-Sprint-2-Statistics/blob/master/module3/Brooke_Drake_DSPT6_U1S2M3_Introduction_to_Bayesian_Inference_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lambda School Data Science Module 123

## Introduction to Bayesian Inference





## Assignment - Code it up!

We used pure math to apply Bayes Theorem to drug tests. Now write Python code to reproduce the results! This is purposefully open ended - you'll have to think about how you should represent probabilities and events. You can and should look things up.

Specific goals/targets:

### 1) Write a function 

`def prob_drunk_given_positive(prob_drunk_prior, false_positive_rate, true_positive_rate):` 

You should only truly need these two values in order to apply Bayes Theorem. In this example, imagine that individuals are taking a breathalyzer test with an 8% false positive rate, a 100% true positive rate, and that our prior belief about drunk driving in the population is 1/1000. 
 - What is the probability that a person is drunk after one positive breathalyzer test?
 - What is the probability that a person is drunk after two positive breathalyzer tests?
 - How many positive breathalyzer tests are needed in order to have a probability that's greater than 95% that a person is drunk beyond the legal limit?

### 2) Explore `scipy.stats.bayes_mvs`  
Read its documentation, and experiment with it on data you've tested in other ways earlier this week.
 - Create a visualization comparing the results of a Bayesian approach to a traditional/frequentist approach. (with a large sample size they should look close to identical, however, take this opportunity to practice visualizing condfidence intervals in general. The following are some potential ways that you could visualize confidence intervals on your graph:
  - [Matplotlib Error Bars](https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.pyplot.errorbar.html)
  - [Seaborn barplot with error bars](https://seaborn.pydata.org/generated/seaborn.barplot.html)
  - [Vertical ines to show bounds of confidence interval](https://www.simplypsychology.org/confidence-interval.jpg)
  - [Confidence Intervals on Box Plots](https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.axes.Axes.boxplot.html)

### 3) In your own words, summarize the difference between Bayesian and Frequentist statistics

If you're unsure where to start, check out [this blog post of Bayes theorem with Python](https://dataconomy.com/2015/02/introduction-to-bayes-theorem-with-python/).



In [3]:
#imports
import numpy as np
import pandas as pd
import scipy.stats as stats
import matplotlib.pyplot as plt
import seaborn as sns

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [0]:
# 1 Define a function

def prob_drunk_given_positive(prob_drunk_prior, false_positive_rate, true_positive_rate):
  prob_not_drunk = 1-prob_drunk_prior
  numerator = (true_positive_rate*prob_drunk_prior)
  marginal = (numerator + (false_positive_rate*prob_not_drunk))
  posterior_probability = numerator/marginal
  return posterior_probability

In [0]:
#given information
prob_drunk_prior = 0.001 # prior belief
P_not_drunk = 1 - prob_drunk_prior  # complement of the prior belief
true_positive_rate = 1 # this is the true positive rate
false_positive_rate = .08 # this is the false positive rate


In [6]:
#What is the probability that a person is drunk after one positive breathalyzer test?
posterior_probability1 = prob_drunk_given_positive(prob_drunk_prior, false_positive_rate, true_positive_rate)
posterior_probability1

0.012357884330202669

In [7]:
#What is the probability that a person is drunk after two positive breathalyzer tests?
posterior_probability2= prob_drunk_given_positive(posterior_probability1, false_positive_rate, true_positive_rate)
posterior_probability2

0.13525210993291495

In [0]:
#How many positive breathalyzer tests are needed in order to have a probability that's greater than 95% that a person is drunk beyond the legal limit?
def prob_drunk_given_positive(prob_drunk_prior, false_positive_rate, true_positive_rate, number_of_tests):
    posterior_probability=prob_drunk_prior
    for test in range(0, number_of_tests):
        prob_drunk_prior = posterior_probability # prior belief
        prob_not_drunk = 1-prob_drunk_prior
        numerator = (true_positive_rate*prob_drunk_prior)
        marginal = (numerator + (false_positive_rate*prob_not_drunk))
        posterior_probability = numerator/marginal
        if posterior_probability > .95:
          print("reached 95% probability on test", test+1)
          break

    return posterior_probability

In [9]:
# How many tests needed for 95% confidence
prob_drunk_given_positive(prob_drunk_prior, false_positive_rate, true_positive_rate, 10) 

reached 95% probability on test 4


0.9606895076105054

# Problem 1, Part Three:
How many positive breathalyzer tests are needed in order to have a probability that's greater than 95% that a person is drunk beyond the legal limit?


* Four tests provide 96% ( >95%) confidence that a person is drunk beyond the legal limit


#Problem 3: In your own words, summarize the difference between Bayesian and Frequentist statistics

* Frequentist statistics are more general and assume no correlation between data points exists. Data is compared with no outside information taken into consideration.

* Bayesian statistics take additional information into account by incorporating observed data and using it to build upon expected outcomes, then comparing that to a pre-formed hypothesis. This more closely identifies how likely a scenario is given that we already know other relevant conditions are true.

## Resources

- [Worked example of Bayes rule calculation](https://en.wikipedia.org/wiki/Bayes'_theorem#Examples) (helpful as it fully breaks out the denominator)
- [Source code for mvsdist in scipy](https://github.com/scipy/scipy/blob/90534919e139d2a81c24bf08341734ff41a3db12/scipy/stats/morestats.py#L139)

## Stretch Goals:

- Go back and study the content from Modules 1 & 2 to make sure that you're really comfortable with them.
- Apply a Bayesian technique to a problem you previously worked (in an assignment or project work) on from a frequentist (standard) perspective
- Check out [PyMC3](https://docs.pymc.io/) (note this goes beyond hypothesis tests into modeling) - read the guides and work through some examples
- Take PyMC3 further - see if you can build something with it!